README
== Synacor Challenge ==

In this challenge, your job is to use this architecture spec to create a
virtual machine capable of running the included binary.  Along the way,
you will find codes; submit these to the challenge website to track
your progress.  Good luck!

== architecture ==

- three storage regions
  - memory with 15-bit address space storing 16-bit values
  - eight registers
  - an unbounded stack which holds individual 16-bit values
- all numbers are unsigned integers 0..32767 (15-bit)
- all math is modulo 32768; 32758 + 15 => 5

== binary format ==

- each number is stored as a 16-bit little-endian pair (low byte, high byte)
- numbers 0..32767 mean a literal value
- numbers 32768..32775 instead mean registers 0..7
- numbers 32776..65535 are invalid
- programs are loaded into memory starting at address 0
- address 0 is the first 16-bit value, address 1 is the second 16-bit value,
  etc

== execution ==

- After an operation is executed, the next instruction to read is immediately
  after the last argument of the current operation.  If a jump was performed,
  the next operation is instead the exact destination of the jump.
- Encountering a register as an operation argument should be taken as reading
  from the register or setting into the register as appropriate.

== hints ==

- Start with operations 0, 19, and 21.
- Here's a code for the challenge website: mxNXGuMXzRII
- The program "9,32768,32769,4,19,32768" occupies six memory addresses and
  should:
  - Store into register 0 the sum of 4 and the value contained in register 1.
  - Output to the terminal the character with the ascii code contained in
    register 0.

== opcode listing ==

halt: 0
  stop execution and terminate the program
set: 1 a b
  set register <a> to the value of <b>
push: 2 a
  push <a> onto the stack
pop: 3 a
  remove the top element from the stack and write it into <a>;
  empty stack = error
eq: 4 a b c
  set <a> to 1 if <b> is equal to <c>; set it to 0 otherwise
gt: 5 a b c
  set <a> to 1 if <b> is greater than <c>; set it to 0 otherwise
jmp: 6 a
  jump to <a>
jt: 7 a b
  if <a> is nonzero, jump to <b>
jf: 8 a b
  if <a> is zero, jump to <b>
add: 9 a b c
  assign into <a> the sum of <b> and <c> (modulo 32768)
mult: 10 a b c
  store into <a> the product of <b> and <c> (modulo 32768)
mod: 11 a b c
  store into <a> the remainder of <b> divided by <c>
and: 12 a b c
  stores into <a> the bitwise and of <b> and <c>
or: 13 a b c
  stores into <a> the bitwise or of <b> and <c>
not: 14 a b
  stores 15-bit bitwise inverse of <b> in <a>
rmem: 15 a b
  read memory at address <b> and write it to <a>
wmem: 16 a b
  write the value from <b> into memory at address <a>
call: 17 a
  write the address of the next instruction to the stack and jump to <a>
ret: 18
  remove the top element from the stack and jump to it; empty stack = halt
out: 19 a
  write the character represented by ascii code <a> to the terminal
in: 20 a
  read a character from the terminal and write its ascii code to <a>; it can be
  assumed that once input starts, it will continue until a newline is
  encountered; this means that you can safely read whole lines from the
  keyboard and trust that they will be fully read
noop: 21
  no operation


CHANGELOG.md
# Revision history for synacor

## 0.1.0.0 -- YYYY-mm-dd

* First version. Released on an unsuspecting world.


src/Main.hs
module Main where

import qualified Control.Monad.State.Lazy as S
import qualified Data.ByteString.Lazy     as B
import qualified Data.Binary.Get          as G
import qualified Data.Map                 as M
import qualified System.Environment       as E

import qualified VM
import qualified VMState                  as V

main :: IO ()
main =
  do bin <- getArgument >>= B.readFile
     S.evalStateT VM.run
       . V.create
       . loadProgramIntoMemory
       . G.runGet bytesToWord16s
       $ bin

getArgument :: IO String
getArgument =
  do args <- E.getArgs
     case args of
       [x] -> pure x
       []  -> error "Specify the filename of the binary as the argument."
       _   -> error "Too many arguments! Only one argument is suppported."

-- | Convert a bytestring to a program (list of little-endian 16-bit integers).
bytesToWord16s :: G.Get [V.Value]
bytesToWord16s =
  do empty <- G.isEmpty
     if empty
        then pure []
        else do x  <- G.getWord16le
                xs <- bytesToWord16s
                pure (x:xs)

-- | Given a program (list of Word16s), create our map of the system memory.
loadProgramIntoMemory :: [V.Value] -> V.Memory
loadProgramIntoMemory = M.fromList . zip [0..]


src/VMState.hs
module VMState where

import qualified Control.Monad.State.Lazy as S
import           Data.Functor                   ((<&>))
import qualified Data.Maybe               as MB
import qualified Data.Map                 as M
import           Data.Word                      (Word16)

type Value        = Word16
type Address      = Value
type Register     = Value
type Memory       = M.Map Address Value

data VM = VM {
    memory  :: Memory
  , stack   :: [Value]
  , pc      :: Address
  , r0      :: Value
  , r1      :: Value
  , r2      :: Value
  , r3      :: Value
  , r4      :: Value
  , r5      :: Value
  , r6      :: Value
  , r7      :: Value
  }

type VMState = S.StateT VM IO

-- | Create a VM with an initial state. The initial memory is passed in.
create :: Memory -> VM
create memory = VM {
    memory  = memory
  , stack   = []
  , pc      = 0
  , r0      = 0
  , r1      = 0
  , r2      = 0
  , r3      = 0
  , r4      = 0
  , r5      = 0
  , r6      = 0
  , r7      = 0
  }

-- | Get the next instructions in the VM from memory, starting from the PC.
nextInstructions :: VMState [Value]
nextInstructions = S.get <&> go
  where
    go :: VM -> [Value]
    go vm = case M.lookup (pc vm) (memory vm) of
              Just x  -> x : go (vm { pc = pc vm + 1})
              Nothing -> []

-- | Evaluate a value as a literal or register value.
getValue :: Value -> VMState Value
getValue x
  | x <= 32767 = pure x
  | otherwise  = readRegister x

-- | Read a value from a memory address.
readMemory :: Address -> VMState Value
readMemory x
  | x <= 32767 = do S.get <&> MB.fromMaybe 0 . M.lookup x . memory
  | otherwise  = error "Invalid memory address (read)."

-- | Set a memory address to a value.
setMemory :: Address -> Value -> VMState ()
setMemory x v
  | x <= 32767 = do vm <- S.get
                    S.put $ vm { memory = M.insert x v $ memory vm }
  | otherwise  = error "Invalid memory address (write)."

-- | Get the current PC.
getPC :: VMState Address
getPC = S.get <&> pc

-- | Increment the program counter by `x` and run the new operation.
incrPC :: Value -> VMState ()
incrPC x =
  do vm <- S.get
     S.put $ vm { pc = pc vm + x }

-- | Set the program counter to `x` and run the new operation.
setPC :: Address -> VMState ()
setPC x =
  do vm <- S.get
     S.put $ vm { pc = x }

-- | Push a value onto the stack.
pushStack :: Value -> VMState ()
pushStack x =
  do vm <- S.get
     S.put $ vm { stack = x : stack vm }

-- | Pop a value off the stack.
popStack :: VMState Value
popStack =
  do vm <- S.get
     case stack vm of
       []   -> error "Cannot pop off empty stack!"
       x:xs -> do S.put $ vm { stack = xs }
                  pure x

-- | Get the value stored in a register.
readRegister :: Register -> VMState Value
readRegister r =
  do vm <- S.get
     case r - 32768 of
       0 -> pure $ r0 vm
       1 -> pure $ r1 vm
       2 -> pure $ r2 vm
       3 -> pure $ r3 vm
       4 -> pure $ r4 vm
       5 -> pure $ r5 vm
       6 -> pure $ r6 vm
       7 -> pure $ r7 vm
       _ -> error $ "Invalid register " <> show r

-- | Set a register to a value.
setRegister :: Register -> Value -> VMState ()
setRegister r v =
  do vm <- S.get
     case r - 32768 of
       0 -> S.put $ vm { r0 = v }
       1 -> S.put $ vm { r1 = v }
       2 -> S.put $ vm { r2 = v }
       3 -> S.put $ vm { r3 = v }
       4 -> S.put $ vm { r4 = v }
       5 -> S.put $ vm { r5 = v }
       6 -> S.put $ vm { r6 = v }
       7 -> S.put $ vm { r7 = v }
       _ -> error $ "Invalid register " <> show r


game/solve_coins.py
#!/usr/bin/env python3

"""
A script to brute force the coin game.

My brain isn't big enough to do the maths.
"""

from collections import namedtuple
import itertools

Coin = namedtuple("Coin", ["color", "value"])

coins = [
    Coin("red", 2),
    Coin("corroded", 3),
    Coin("shiny", 5),
    Coin("concave", 7),
    Coin("blue", 9),
]


def check_perm(p):
    s = p[0].value + p[1].value * (p[2].value ** 2) + (p[3].value ** 3) - p[4].value
    return s == 399


def take_coins(p):
    return "".join(f"take {c.color} coin\n" for c in p)


def use_coins(p):
    return "".join(f"use {c.color} coin\n" for c in p)


for perm in itertools.permutations(coins, r=5):
    if check_perm(perm):
        print(take_coins(perm))
        print(use_coins(perm))
        break
else:
    print("No configuration found... you suck at programming.")


src/VM.hs
module VM where

import qualified Control.Monad            as M
import qualified Control.Monad.State.Lazy as S
import           Data.Bits                      ((.&.), (.|.), xor)
import qualified Data.Char                as C

import qualified VMState                  as V

-- | Execute a program in our VM.
run :: V.VMState ()
run =
  do ops <- V.nextInstructions
     case ops of
       (0 :_   ) -> pure ()
       (1 :args) -> set  args >> continue 3
       (2 :args) -> push args >> continue 2
       (3 :args) -> pop  args >> continue 2
       (4 :args) -> eq   args >> continue 4
       (5 :args) -> gt   args >> continue 4
       (6 :args) -> jmp  args >> continue 0
       (7 :args) -> jt   args >> continue 0
       (8 :args) -> jf   args >> continue 0
       (9 :args) -> add  args >> continue 4
       (10:args) -> mult args >> continue 4
       (11:args) -> mod' args >> continue 4
       (12:args) -> and' args >> continue 4
       (13:args) -> or'  args >> continue 4
       (14:args) -> not' args >> continue 3
       (15:args) -> rmem args >> continue 3
       (16:args) -> wmem args >> continue 3
       (17:args) -> call args >> continue 0
       (18:args) -> ret  args >> continue 0
       (19:args) -> out  args >> continue 2
       (20:args) -> in'  args >> continue 2
       (21:args) -> noop args >> continue 1
       (x :xs  ) -> error $ "Found unknown opcode "
                            <> show x
                            <> " with arguments "
                            <> show (take 3 xs)
       []        -> error "Instruction list is empty but program has not halted!"

-- | Increment the PC by the provided value and continue execution.
continue :: V.Value -> V.VMState ()
continue x = V.incrPC x >> run

-- | Opcode 1:  Set register.
set :: [V.Value] -> V.VMState ()
set (x:y:_) = V.getValue y >>= V.setRegister x

-- | Opcode 2:  Push stack.
push :: [V.Value] -> V.VMState ()
push (x:_) = V.getValue x >>= V.pushStack

-- | Opcode 3:  Pop stack.
pop :: [V.Value] -> V.VMState ()
pop (x:_) = V.popStack >>= V.setRegister x

-- | Opcode 4:  EQ comparison.
eq :: [V.Value] -> V.VMState ()
eq (x:y:z:_) =
  do a1 <- V.getValue y
     a2 <- V.getValue z
     V.setRegister x (if a1 == a2 then 1 else 0)

-- | Opcode 5:  GT comparison.
gt :: [V.Value] -> V.VMState ()
gt (x:y:z:_) =
  do a1 <- V.getValue y
     a2 <- V.getValue z
     V.setRegister x (if a1 > a2 then 1 else 0)

-- | Opcode 6:  Jump.
jmp :: [V.Value] -> V.VMState ()
jmp (x:_) = V.getValue x >>= V.setPC

-- | Opcode 7:  Jump non-zero.
jt :: [V.Value] -> V.VMState ()
jt (x:y:_) =
  do val <- V.getValue x
     if val == 0
        then V.incrPC 3
        else V.setPC y

-- | Opcode 8:  Jump zero.
jf :: [V.Value] -> V.VMState ()
jf (x:y:_) =
  do val <- V.getValue x
     if val == 0
        then V.setPC y
        else V.incrPC 3

-- | Opcode 9:  Addition.
add :: [V.Value] -> V.VMState ()
add (x:y:z:_) =
  do a1 <- V.getValue y
     a2 <- V.getValue z
     V.setRegister x $ (a1 + a2) `mod` 32768

-- | Opcode 10: Multiplication.
mult :: [V.Value] -> V.VMState ()
mult (x:y:z:_) =
  do a1 <- V.getValue y
     a2 <- V.getValue z
     V.setRegister x $ (a1 * a2) `mod` 32768

-- | Opcode 11: Mod/Remainder.
mod' :: [V.Value] -> V.VMState ()
mod' (x:y:z:_) =
  do a1 <- V.getValue y
     a2 <- V.getValue z
     V.setRegister x $ (a1 `rem` a2) `mod` 32768

-- | Opcode 12: Bitwise and operator.
and' :: [V.Value] -> V.VMState ()
and' (x:y:z:_) =
  do a1 <- V.getValue y
     a2 <- V.getValue z
     V.setRegister x (a1 .&. a2)

-- | Opcode 13: Bitwise or operator.
or' :: [V.Value] -> V.VMState ()
or' (x:y:z:_) =
  do a1 <- V.getValue y
     a2 <- V.getValue z
     V.setRegister x (a1 .|. a2)

-- | Opcode 14: Bitwise not/inverse operator.
not' :: [V.Value] -> V.VMState ()
not' (x:y:_) = V.getValue y >>= V.setRegister x . (`xor` 0x7FFF)

-- | Opcode 15: Read memory from 2nd arg and write to 1st arg.
rmem :: [V.Value] -> V.VMState ()
rmem (x:y:_) = V.getValue y >>= V.readMemory >>= V.setRegister x

-- | Opcode 16: Get value of 2nd arg and write to memory at 1st arg.
wmem :: [V.Value] -> V.VMState ()
wmem (x:y:_) = M.join $ V.setMemory <$> V.getValue x <*> V.getValue y

-- | Opcode 17: Call.
call :: [V.Value] -> V.VMState ()
call (x:_) =
  do V.getPC >>= V.pushStack . (+2)
     V.getValue x >>= V.setPC

-- | Opcode 18: Return.
ret :: [V.Value] -> V.VMState ()
ret _ = V.popStack >>= V.setPC

-- | Opcode 19: Print char.
out :: [V.Value] -> V.VMState ()
out (x:_) = V.getValue x >>= S.liftIO . putChar . C.chr . fromIntegral

-- | Opcode 20: Input.
in' :: [V.Value] -> V.VMState ()
in' (x:_) = S.liftIO getChar >>= V.setRegister x . fromIntegral . C.ord

-- | Opcode 21: No operation.
noop :: [V.Value] -> V.VMState ()
noop _ = pure ()
